pemFioi/quickpi/blocklyQuickPi_lib.js

```diff
@@ -8826,10 +8832,65 @@ var getContext = function (display, infos, curLevel) {
         context.runner.waitDelay(callback, millis);
     };
 
-    context.quickpi.getTemperature = function(location, callback) {
-        var retVal =  25;
 
-        context.waitDelay(callback, retVal);
+    // TODO: change url to the "prod" one
+    context.quickpi.getTemperatureFromCloudUrl = "https://mapadev.com/nicolastest/weather.php";
+
+    // setup the supported towns
+    $.get(context.quickpi.getTemperatureFromCloudUrl + "?q=" + "supportedtowns", function(towns) {
+        context.quickpi.getTemperatureFromCloudSupportedTowns = towns;
+    });
+
+    // We create a cache so there is less calls to the api and we get the results of the temperature faster
+    context.quickpi.getTemperatureFromCloudCache = [];
+
+    context.quickpi.getTemperatureFromCloud = function(location, callback) {
+
+        function _updateGetTemperatureFromCloudCache(cache, location, newVal) {
+            for (var i = 0; i < cache.length; i++) {
+                var curr = cache[i];
+                if (curr.location === location) {
+                    curr.lastUpdate = Date.now();
+                    curr.temperature = newVal;
+                    return;
+                }
+            }
+            cache.push({
+                location: location,
+                lastUpdate: Date.now(),
+                temperature: newVal
+            });
+        }
+
+        var url = context.quickpi.getTemperatureFromCloudUrl;
+
+        if (!context.quickpi.getTemperatureFromCloudSupportedTowns.includes(location))
+            throw strings.messages.getTemperatureFromCloudWrongValue.format(location);
+
+        var cache = context.quickpi.getTemperatureFromCloudCache;
+        for (var i = 0; i < cache.length; i++) {
+            var curr = cache[i];
+            if (curr.location === location) {
+                // if last update was less than 10 min
+                if (((Date.now() - curr.lastUpdate) / 1000) / 60 < 10) {
+                    context.waitDelay(callback, curr.temperature);
+                    return;
+                }
+                break;
+            }
+        }
+
+        var cb = context.runner.waitCallback(callback);
+        $.get(url + "?q=" + location, function(data) {
+            // If the server return invalid it mean that the town given is not supported
+            if (data === "invalid") {
+                cb(0);
```

Comment:
I wouldn't return `0` in that case. Maybe throw an exception?

Assuming that if a town is invalid, it will always be, then it's better to tell the user that something is wrong with a proper exception, rather than silently return zero.

Reply:
Thats the first behavior I tried to do (throw something there), but it was not possible (the throw after the function "context.runner.waitCallback" is not catch by the runner and does not display a good error message).

"invalid" is returned by the php server only when the user is trying to get an unsupported town.

So I decided to check if the town is supported before at lign 8850 of last commit and throw the exception there if not. So the return 0 is never supposed to happen.

The only case it can happen is this:
When the user load his page, he gets Paris and Madrid as supported towns.
After the load someone of this company decides to remove "Madrid" from the supported towns in the php.
The user try to access at Madrid, but it was deleted from supported towns, the server returns "invalid" so we return 0.

But that is a very rare case moreover no one would have any interest of suppressing any supported town I think

Reply:
Ok, that makes sense. Add a comment above the line `cb(0);` to write all that, so that we don't forget!

Reply:
A comment has been added to explain it